### PR TITLE
APP-329: [C/C++] Distinguish between CMake variables and compile definitions in cmakegen

### PR DESCRIPTION
**Pull request by Slash32**
See #354 for the original PR.

### CODE REVIEW

File: scripts/cmakegen/cmakegen.py [hunk 1]
```diff
@@ -39,12 +39,13 @@ def __init__(self, features):
     def generate(self, language):
         if self._features is None:
             return str()
-        template_c = 'set(WITH_EXTENSION_%extension OFF)\n'
-        template_cpp = 'set(KAA_WITHOUT_%extension 1)\n'
+        template_c = '\nset(WITH_EXTENSION_%extension OFF)'
+        template_cpp = '\nset(KAA_WITHOUT_%extension 1)'
+        comment = '# Disable unused features' if len(self._features) < 4 else str()
```
Comment (rasendubi) [inline scripts/cmakegen/cmakegen.py]: Why?

Also, the condition seems to be flipped. Are you emitting comment when there are 0-3 disabled features? (I would do if there are 1+ of them.)

Comment (rasendubi) [inline scripts/cmakegen/cmakegen.py]: `str()` -> `""`

File: scripts/cmakegen/cmakegen.py [hunk 2]
```diff
@@ -56,18 +57,25 @@ def _generate_from_template(self, template):
...
-class DefinitionsGenerator:
-    def __init__(self, definitions):
-        self._definitions = dict(definitions) if definitions is not None else dict()
+    def add_variables(self, **kwargs):
+        self._variables.update(kwargs)
 
-    def add_definitions(self, **kwargs):
-        self._definitions.update(kwargs)
+    def generate(self):
+        out = '# Set configuration variables\n' if len(self._variables) > 0 else str()
```
Comment (rasendubi) [inline scripts/cmakegen/cmakegen.py]: `str()` -> `""`

File: scripts/cmakegen/cmakegen.py [hunk 3]
```diff
@@ -56,18 +57,25 @@ def _generate_from_template(self, template):
...
+        for k, v in self._variables.iteritems():
+            out = out + CMakeGenTemplate(VARIABLE_TEMPLATE).safe_substitute(variable=k, value=v)
+        return out
 
+class DefinitionsGenerator(VariablesGenerator):
     def generate(self):
-        out = str()
-        for k, v in self._definitions.iteritems():
-            out = out + CMakeGenTemplate(DEFINITION_TEMPLATE).safe_substitute(variable=k, value=v)
+        out = '# Set compile definitions\n' if len(self._variables) > 0 else str()
```
Comment (rasendubi) [inline scripts/cmakegen/cmakegen.py]: `str()` -> `""`

File: scripts/cmakegen/config.yaml [hunk 4]
```diff
@@ -36,15 +36,15 @@ datacollectiondemo:
 
 eventsdemo:
     appname: 'EventsDemo'
-    languages: ['c', 'cpp']
+    languages: ['cpp']
```
Comment (rasendubi) [inline scripts/cmakegen/config.yaml:39]: Doesn't look like this belongs to the issue.

File: configurationdemo/source/c/CMakeLists.txt [hunk 5]
```diff
@@ -22,26 +22,19 @@ set(WITH_EXTENSION_NOTIFICATIONS OFF)
...
-if (NOT DEFINED KAA_TARGET)
-    set(KAA_TARGET "posix")
-endif (NOT DEFINED KAA_TARGET)
-add_definitions(-DKAA_TARGET="${KAA_TARGET}")
-
-
-
 if (NOT DEFINED KAA_MAX_LOG_LEVEL)
-set(KAA_MAX_LOG_LEVEL 3)
+    set(KAA_MAX_LOG_LEVEL "3")
```
Comment (rasendubi) [inline configurationdemo/source/c/CMakeLists.txt]: Again, why that's a string?

**Review by rasendubi — COMMENTED** [review]
Looks ok to me. Though, my original comment on inventing the wheel still applies.

File: scripts/cmakegen/cmakegen.py (hunk 1 above)
Comment (RostakaGmfun) [reply scripts/cmakegen/cmakegen.py]: `self._features` variable contains a list of _enabled_ features

Comment (rasendubi) [reply scripts/cmakegen/cmakegen.py]: so... why 4?

**Comment by Slash32** [comment]
@kaa-leeroy test this fully please

**Pull request merged**